torch/csrc/jit/script/compiler.cpp

```diff
@@ -418,45 +421,57 @@ struct to_ir {
-    // https://github.com/onnx/onnx/blob/master/docs/Operators.md#experimental-loop
-    // TODO: implement scan_outputs
-
-    // the format of the Loop instruction is:
-    // loop_carried_outputs* = Loop(max_trip_count, start_condition, loop_carried_inputs*)
-    //                          block0(loop_counter, loop_carried_block*) {
-    //                             <body>
-    //                             -> (continue_condition, loop_carried_block_outputs*)
-    //                          }
-    // all loop_carried_... lists are the same length and represent the value of
-    // loop-carried variables whose definitions are updated as the loop executes
-    // in a way that ensure single static assignment.
-
-    // TODO: clarify that this is an optional input that isn't needed here
-    Value* max_trip_count_dummy = emitConst(stmt.range(), INT_MAX, "i")[0];
-    Value* cond_value = emitExpr(stmt.cond(), 1)[0];
-
+  // *********************** Loop Operators ************************************
+  // Emits a loop operators conforming to the semantics specified at
+  // https://github.com/onnx/onnx/blob/master/docs/Operators.md#experimental-loop
+  // TODO: implement scan_outputs
+
+  // the format of the Loop instruction is:
+  // loop_carried_outputs* = Loop(max_trip_count, start_condition,
+  // loop_carried_inputs*)
+  //                          block0(loop_counter, loop_carried_block*) {
+  //                             <body>
+  //                             -> (continue_condition,
+  //                             loop_carried_block_outputs*)
+  //                          }
+  // all loop_carried_... lists are the same length and represent the value of
+  // loop-carried variables whose definitions are updated as the loop executes
+  // in a way that ensure single static assignment.
+
+  void emitLoopCommon(
+      Value* max_trip_count,
+      Value* cond_value,
+      const List<Stmt>& body,
+      const Stmt& stmt,
+      const Expr* cond,
+      const std::vector<std::string>& itr_idents) {
     Node* n = graph->insertNode(create(kLoop, stmt.range(), 0));
-    n->addInput(max_trip_count_dummy);
+    Value* trip_count = n->addInput(max_trip_count);
     n->addInput(cond_value);
     auto* body_block = n->addBlock();
-    // Trip count required by spec. Since this is a while loop, we do not
-    // provide access to this from user code
-    // TODO: it seems like we should implement a `for` loop as well, otherwise
-    // we'll probably have to pattern match iteration number machinery in user
-    // code to conform to the spec
     body_block->addInput(); // Iteration num
     size_t skip_inputs_num = 1;
 
     {
       pushFrame(body_block);
+      if (itr_idents.size() > 1) {
+        throw ErrorReport(stmt)
+            << "Only one iteration variable allowed right now.";
+      }
+      for (size_t i = 0; i < itr_idents.size(); ++i) {
+        environment_stack->setVar(itr_idents[i], trip_count);
```

Comment:
You've just asserted that there's a single iterator so this loop is unnecessary and confusing